lib/charms/loki_k8s/v0/log_proxy.py

```diff
@@ -0,0 +1,589 @@
+            clients = [c for c in clients if c != {"url": agent_url}]
+            current_config["clients"] = clients
+            return current_config
+
+        return current_config
+
+    def _server_config(self) -> dict:
+        """Generates the server section of the Promtail config file.
+
+        Returns:
+            A dict representing the `server` section.
+        """
+        return {
+            "server": {
+                "http_listen_port": HTTP_LISTEN_PORT,
+                "grpc_listen_port": GRPC_LISTEN_PORT,
+            }
+        }
+
+    def _positions(self) -> dict:
+        """Generates the positions section of the Promtail config file.
+
+        Returns:
+            A dict representing the `positions` section.
+        """
+        return {"positions": {"filename": WORKLOAD_POSITIONS_PATH}}
+
+    def _scrape_configs(self) -> dict:
+        """Generates the scrape_configs section of the Promtail config file.
+
+        Returns:
+            A dict representing the `scrape_configs` section.
+        """
+        # TODO: use the JujuTopology object
+        config = {
+            "targets": ["localhost"],
+            "labels": {
+                "job": "juju_{}_{}_{}".format(
+                    self._charm.model.name,
+                    self._charm.model.uuid,
+                    self._charm.model.app.name,
+                ),
+                "__path__": "",
+            },
+        }
+
+        return {
+            "scrape_configs": [
+                {
+                    "job_name": "system",
+                    "static_configs": self._generate_static_configs(config),
+                }
+            ]
+        }
+
+    def _generate_static_configs(self, config: dict) -> list:
+        """Generates static_configs section.
+
+        Returns:
+            - a list of dictionaries representing static_configs section
+        """
+        static_configs = []
+
+        for _file in self._log_files:
+            conf = deepcopy(config)
+            conf["labels"]["__path__"] = _file
+            static_configs.append(conf)
+
+        return static_configs
+
+
+class LogProxyProvider(RelationManagerBase):
+    """LogProxyProvider class."""
+
+    def __init__(self, charm, relation_name: str = DEFAULT_RELATION_NAME):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.framework.observe(
+            self._charm.on.log_proxy_relation_changed, self._on_log_proxy_relation_changed
+        )
+
+    def _on_log_proxy_relation_changed(self, event):
+        if event.relation.data[self._charm.unit].get("data") is None:
+            data = {}
+            data.update(json.loads(self._loki_push_api))
+            data.update(json.loads(self._promtail_binary_url))
+            event.relation.data[self._charm.unit].update({"data": json.dumps(data)})
+
+    @property
+    def _promtail_binary_url(self) -> str:
```

Comment:
Not sure why the download url needs to travel over relation data rather than being hard-coded there instead of here.

Reply:
Because the Provider side maintain this info up to date.